File: Programme_Python/fichier/exo-3.py
import os
import struct

def ecriture_binaire(fichier, n):
    tab = []

    # Saisie des elements
    print(f"Entrez {n} entiers :")
    for i in range(n):
        while True:
            try:
                val = int(input(f"==> "))
                tab.append(val)
                break
            except ValueError:
                print("Entrée invalide. Veuillez entrer un entier.")

    # ouverture du fichier en mode binaire
    try:
        with open(fichier, "wb") as f:
            # On écrit les entiers en binaire
            for val in tab:
                # 'i' = int (4 octets), format standard C
                f.write(struct.pack('i', val))
    except OSError as e:
        print(f"Erreur ouverture ou écriture fichier : {e}")
        return

    # Définition des permissions : écriture pour le groupe, lecture pour les autres
    # 0o620 = rw--w---- (groupe write, autres read) ; adapté de l'exemple C
    # Si tu veux exactement comme C : rw------- (owner) | rw- (group) | r-- (others) => 0o764
    try:
        os.chmod(fichier, 0o624)  # rw- r-- -w-
    except OSError as e:
        print(f"Erreur chmod fichier : {e}")
        return

    print(f"Tableau sauvegardé dans '{fichier}' avec les permissions correctes.")


if __name__ == "__main__":
    try:
        n = int(input("Entrez le nombre d'entiers du tableau : "))
        if n <= 0:
            print("Nombre invalide.")
        else:
            ecriture_binaire("tableau.bin", n)
    except ValueError:
        print("Entrée invalide.")

File: Programme_C/chapitre-6/exo3.c
/*
  Écrire un programme qui saisit au clavier un tableau d’entiers et sauve-
 garde ce tableau au format binaire dans un fichier ayant permission en écriture pour le groupe
 du fichier et en lecture seule pour les autres utilisateurs.
*/

#include <stdio.h>
#include <stdlib.h>
#include <fcntl.h> 
#include <unistd.h>


// Fonction saisit un tableau et le sauvegarde en binaire avec les permissions

void ecriture_binaire(const char *fichier, int n) 
{
     int *tab = malloc(n * sizeof(int));

     if (tab == NULL) 
     {
        perror("Erreur d'allocation memoire");
        return;
     }

     printf("Entrez %d entiers :\n", n);
     for (int i = 0; i < n; i++) 
     {
         printf("==> ") ;
         if (scanf("%d", &tab[i]) != 1) 
         {
             printf("Entree invalide.\n");
             free(tab);
             return;
         }
     }

     int fd = open(fichier, O_WRONLY | O_CREAT | O_TRUNC , S_IWUSR | S_IRUSR | S_IWGRP | S_IROTH);
     if (fd == -1) 
     {
        perror("Erreur ouverture fichier");
        free(tab);
        return;
     }

     
     ssize_t bytes = write(fd , tab , sizeof(int) * n ) ;
     if(bytes != (ssize_t) (sizeof(int) * n) )
     {
        perror("Erreur ecriture fichier");
        close(fd);
        free(tab);
        return;
     }

     close(fd);
     free(tab);


     printf("Tableau sauvegarde dans '%s' avec les permissions correctes.\n", fichier);

}

int main() 
{
     int n = 0 ;
     char nom_fichier[] = "tableau.bin" ;

     printf("Entrez le nombre d'entiers du tableau : ");
     if (scanf("%d", &n) != 1 || n <= 0) 
     {
         printf("Nombre invalide.\n");
         return 1;
     }

  
     ecriture_binaire(nom_fichier, n);

     return 0;
}